### PR TITLE
Handle BQ write result as ClosedTap side output

### PR DESCRIPTION
**Pull request by RustedBones**
Proposal to address #5103

This potentially could also be an alternative solution for #4995

### CODE REVIEW

File: scio-google-cloud-platform/src/main/scala/com/spotify/scio/bigquery/BigQueryIO.scala [hunk 1]
```diff
@@ -402,10 +401,42 @@ final case class BigQueryTypedTable[T: Coder](
...
       .pipe(w => Option(params.failedInsertRetryPolicy).fold(w)(w.withFailedInsertRetryPolicy))
-      .pipe(params.configOverride)
-      .pipe(w => if (params.extendedErrorInfo == Disabled) w else w.withExtendedErrorInfo())
+      .pipe(w => Option(params.configOverride).fold(w)(_(w)))
+      .pipe(w => if (params.extendedErrorInfo) w.withExtendedErrorInfo() else w)
+
+    val wr = data.applyInternal(transform)
+
+    var tuple = PCollectionTuple.empty(data.context.pipeline)
+    if (
```
Comment (clairemcginty) [inline scio-google-cloud-platform/src/main/scala/com/spotify/scio/bigquery/BigQueryIO.scala:410]: 🎨  might be more readable as a `match` statement on `(params.method, data.internal.isBounded` - wdyt?

File: scio-google-cloud-platform/src/main/scala/com/spotify/scio/bigquery/BigQueryIO.scala [hunk 2]
```diff
@@ -402,10 +401,42 @@ final case class BigQueryTypedTable[T: Coder](
...
+      (params.method == WriteMethod.DEFAULT && data.internal.isBounded == PCollection.IsBounded.UNBOUNDED)
+    ) {
+      tuple = tuple.and(BigQueryIO.SuccessfulInserts.tupleTag, wr.getSuccessfulInserts)
+    } else if (BigQueryIO.storageWriteMethod.contains(params.method)) {
+      tuple = tuple.and(
+        BigQueryIO.SuccessfulStorageApiInserts.tupleTag,
+        wr.getSuccessfulStorageApiInserts
+      )
+    } else {
+      throw new IllegalArgumentException(s"Unknown write method: ${params.method}")
```
Comment (clairemcginty) [inline scio-google-cloud-platform/src/main/scala/com/spotify/scio/bigquery/BigQueryIO.scala:426]: imo, this seems like more of a case of an `UnsupportedOperationException` than illegal/unknown argument exception

**Review by clairemcginty — COMMENTED** [review]
I really like the idea of enclosing the PCollectionTuple in `Tap`! Makes sense and it simplifies our BQ config a lot 👍

File: scio-core/src/main/scala/com/spotify/scio/io/ScioIO.scala [hunk 3]
```diff
@@ -167,6 +169,27 @@ trait KeyedIO[T] { self: ScioIO[T] =>
   def keyCoder: Coder[KeyT]
 }
 
+trait WriteResultIO[T] { self: ScioIO[T] =>
+  override private[scio] def writeWithContext(
+    data: SCollection[T],
+    params: WriteP
+  ): ClosedTap[tapT.T] = {
+    val (tap, outputs) =
+      if (data.context.isTest) writeWithResult(data, params) else writeWithResult(data, params)
```
Comment (RustedBones) [inline scio-core/src/main/scala/com/spotify/scio/io/ScioIO.scala:178]: TODO handle testing

**Review by clairemcginty — APPROVED** [review]

**Pull request merged**